src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -196,8 +199,27 @@ private static MethodSymbol GetEntryPoint(CSharpCompilation compilation, PEModul
             diagnostics.AddRange(entryPointAndDiagnostics.Diagnostics);
 
             var entryPoint = entryPointAndDiagnostics.MethodSymbol;
-            var synthesizedEntryPoint = entryPoint as SynthesizedEntryPointSymbol;
-            if (((object)synthesizedEntryPoint != null) &&
+
+            if (entryPoint == null)
+            {
+                Debug.Assert(entryPointAndDiagnostics.Diagnostics.HasAnyErrors() || !compilation.Options.Errors.IsDefaultOrEmpty);
+                return null;
+            }
+
+            SynthesizedEntryPointSymbol synthesizedEntryPoint = null;
+            bool addedDefinition = false;
+            if (entryPoint is SynthesizedEntryPointSymbol s)
+            {
+                synthesizedEntryPoint = s;
+            }
+            else if (entryPoint.HasAsyncMainReturnType(compilation) && compilation.LanguageVersion >= LanguageVersion.CSharp7_1)
+            {
+                synthesizedEntryPoint = new AsyncForwardEntryPoint(compilation, diagnostics, entryPoint.ContainingType, entryPoint);
```

Comment:
It seems odd that we are attaching the `diagnostics` to only the async sythesized method symbol.  What is the rationale for this?  

In the case the `diagnostics` parameter had existing diagnostics in it, possible for a bag, it would be attaching those errors to the generated main as well which seems incorrect.  As say using `entryPointAndDiagnostics.Diagnostics`.  Even that though feels like we are double tracking diagnostics here.